Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm

```diff
@@ -0,0 +1,825 @@
+    auto& deniedPermissionMatchPatterns = this->deniedPermissionMatchPatterns();
+
+    // If the cache still has the URL, then it has not expired.
+    if (m_cachedPermissionURLs.contains(coreURL)) {
+        PermissionState cachedState = m_cachedPermissionStates.get(coreURL);
+
+        // We only want to return an unknown cached state if the SkippingRequestedPermissions option isn't used.
+        if (cachedState != PermissionState::Unknown || skipRequestedPermissions) {
+            // Move the URL to the end, so it stays in the cache longer as a recent hit.
+            m_cachedPermissionURLs.appendOrMoveToLast(coreURL);
+
+            if ((cachedState == PermissionState::RequestedExplicitly || cachedState == PermissionState::RequestedImplicitly) && skipRequestedPermissions)
+                return PermissionState::Unknown;
+
+            return cachedState;
+        }
+    }
+
+    auto cacheResultAndReturn = ^PermissionState(PermissionState result) {
+        m_cachedPermissionURLs.appendOrMoveToLast(coreURL);
+        m_cachedPermissionStates.set(coreURL, result);
+
+        ASSERT(m_cachedPermissionURLs.size() == m_cachedPermissionURLs.size());
+
+        if (m_cachedPermissionURLs.size() <= maximumCachedPermissionResults)
+            return result;
+
+        URL firstCachedURL = m_cachedPermissionURLs.takeFirst();
+        m_cachedPermissionStates.remove(firstCachedURL);
+
+        ASSERT(m_cachedPermissionURLs.size() == m_cachedPermissionURLs.size());
+
+        return result;
+    };
+
+    // First, check for patterns that are specific to certain domains, ignoring wildcard host patterns that
+    // match all hosts. The order is denied, then granted. This makes sure denied takes precedence over granted.
+
+    auto urlMatchesPatternIgnoringWildcardHostPatterns = ^(WebExtensionMatchPattern& pattern) {
+        if (pattern.matchesAllHosts())
+            return false;
+        return pattern.matchesURL(url);
+    };
+
+    for (auto& deniedPermissionEntry : deniedPermissionMatchPatterns) {
+        if (urlMatchesPatternIgnoringWildcardHostPatterns(deniedPermissionEntry.key))
+            return cacheResultAndReturn(PermissionState::DeniedExplicitly);
+    }
+
+    for (auto& grantedPermissionEntry : grantedPermissionMatchPatterns) {
+        if (urlMatchesPatternIgnoringWildcardHostPatterns(grantedPermissionEntry.key))
+            return cacheResultAndReturn(PermissionState::GrantedExplicitly);
+    }
+
+    // Next, check for patterns that are wildcard host patterns that match all hosts (<all_urls>, *://*/*, etc),
+    // also checked in denied, then granted order. Doing these wildcard patterns separately allows for blanket
+    // patterns to be set as default policies while allowing for specific domains to still be granted or denied.
+
+    auto urlMatchesWildcardHostPatterns = ^(WebExtensionMatchPattern& pattern) {
+        if (!pattern.matchesAllHosts())
+            return false;
+        return pattern.matchesURL(url);
```

Comment:
Is there a time where `pattern.matchesURL` would be false if the pattern matches all hosts?

Reply:
If the pattern is "http://*/*" and the scheme or path don't match still.